service/src/main/java/bio/terra/pipelines/service/ImputationService.java

```diff
@@ -1,40 +1,140 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataIntegrityViolationException;
 import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
 
 /** Service to encapsulate logic used to run an imputation pipeline */
 @Service
 public class ImputationService {
   private static final Logger logger = LoggerFactory.getLogger(ImputationService.class);
+  private final ImputationJobsRepository imputationJobsRepository;
+  private final PipelineInputsRepository pipelineInputsRepository;
   private LeonardoService leonardoService;
   private SamService samService;
   private WdsService wdsService;
+  private final JobService jobService;
   private ImputationConfiguration imputationConfiguration;
 
   @Autowired
   ImputationService(
+      ImputationJobsRepository imputationJobsRepository,
+      PipelineInputsRepository pipelineInputsRepository,
       LeonardoService leonardoService,
       SamService samService,
       WdsService wdsService,
+      JobService jobService,
       ImputationConfiguration imputationConfiguration) {
+    this.imputationJobsRepository = imputationJobsRepository;
+    this.pipelineInputsRepository = pipelineInputsRepository;
     this.leonardoService = leonardoService;
     this.samService = samService;
     this.wdsService = wdsService;
+    this.jobService = jobService;
     this.imputationConfiguration = imputationConfiguration;
   }
 
+  /**
+   * Creates a new Imputation pipeline service job, using a Stairway flight, based on a user's
+   * request. Returns jobId of new job (which is the same as the flightId) if flight submission is
+   * successful.
+   *
+   * @param userId
+   * @param pipelineVersion
+   * @return String jobId
+   *     <p>Note that the information in the requested job will grow over time, along with the
+   *     following related classes:
+   * @see ImputationJob
+   */
+  @Transactional
+  public UUID createImputationJob(
+      String userId, String description, String pipelineVersion, Object pipelineInputs) {
+    logger.info("Create new imputation version {} job for user {}", pipelineVersion, userId);
+
+    JobBuilder jobBuilder =
+        jobService
+            .newJob()
+            .jobId(createJobId())
+            .flightClass(RunImputationJobFlight.class)
+            .addParameter(JobMapKeys.PIPELINE_ID.getKeyName(), PipelinesEnum.IMPUTATION)
+            .addParameter(JobMapKeys.USER_ID.getKeyName(), userId)
+            .addParameter(JobMapKeys.DESCRIPTION.getKeyName(), description)
+            .addParameter(RunImputationJobFlightMapKeys.PIPELINE_VERSION, pipelineVersion)
+            .addParameter(RunImputationJobFlightMapKeys.PIPELINE_INPUTS, pipelineInputs);
+
+    return jobBuilder.submit();
+  }
+
+  // only used in tests
+  @VisibleForTesting
+  public ImputationJob getImputationJob(UUID jobId, String userId) {
```

Comment:
do we need this function and the one below it if theyre only used for testing?  cant we just call the imputationJobsRepository in the tests directly?

Reply:
:facepalm: sure can, thanks!